### PR TITLE
Reference lang items during AST lowering

### LINKED ISSUE #60607: Preallocate `DefId`s for lang items and use lang items in lowering
Right now, our lowering code uses string-paths to call e.g. `Iterator::next` when lowering `for` loops:

https://github.com/rust-lang/rust/blob/c6ac57564852cb6e2d0db60f7b46d9eb98d4b449/src/librustc/hir/lowering.rs#L4653-L4655

This triggers a name resolution call

https://github.com/rust-lang/rust/blob/c6ac57564852cb6e2d0db60f7b46d9eb98d4b449/src/librustc/hir/lowering.rs#L5263-L5273

which will then give us a resolved path, but that's

1. potentially expensive, since it happens *every time* a specific path is generated (which happens a lot for `?` and `for` expansions/lowerings)
2. seems uncool since it's stringly typed

I think it would be neat if we could instead create the `lang_items` table early in the process and just grab the appropriate item out of that table.

cc @Centril @eddyb @petrochenkov 

As a first experiment to check out the impact we could create a cache for path resolutions so we only run the path resolution code once and just overwrite all `HirId`s in the generated path with new `HirId`s whenever we fetch a cached resolution.

### LINKED ISSUE #61019: `extern crate ... as std` overrides the injected crate on 2018 edition
On the 2018 edition the following code:

```rust
extern crate some_crate as std; // or as core with #![no_std]
```

will:

* ~~Try to resolve `prelude::v1::*` in `some_crate`, failing if it doesn't exist. This probably limits the plausible cases where this could happen to `extern crate core as std`, and `extern crate core as std` with `#![no_std]`.~~
* Use `some_crate` in ~~any built-in derives and for~~ HIR lowering.

I would expect that the prelude and derives/HIR lowering to always use the injected std/core. Paths starting with `std` should still resolve to the in the crate the user provided.

In the 2015 edition it creates a name conflict with the implicit `extern crate std`.

cc @petrochenkov @eddyb

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #60607 and fixes #61019.

This PR introduces `QPath::LangItem` to the HIR and uses it in AST lowering instead of constructing a `hir::Path` from a slice of symbols:

- Credit for much of this work goes to @matthewjasper, I basically just [rebased their earlier work](https://github.com/matthewjasper/rust/commit/a227c706b7809ff07021baf3856b7540d5b57f8a#diff-c0f791ead38d2d02916faaad0f56f41d).
- ~~Changes to Clippy might not be correct, they compile but attempting to run tests through `./x.py` produced failures which appeared spurious, so I didn't run any clippy tests.~~
- Changes to save analysis might not be correct - tests pass but I don't have a lot of confidence in those changes being correct.
- I've used `GenericBounds::LangItemTrait` rather than changing `PolyTraitRef`, as suggested by @matthewjasper [in this comment](https://github.com/matthewjasper/rust/commit/a227c706b7809ff07021baf3856b7540d5b57f8a#r40107992) but I'd prefer that be left for a follow-up.
- I've split things into smaller commits fairly arbitrarily to make the diff easier to review, each commit should compile but might not pass tests until the final commit.

r? @oli-obk 
cc @matthewjasper

### CODE REVIEW

**Comment by petrochenkov** [comment]
(I'll review this next weekend or during the week.)

File: src/librustc_hir/hir.rs [hunk 1]
```diff
@@ -1677,6 +1681,35 @@ pub enum QPath<'hir> {
...
+impl<'hir> QPath<'hir> {
+    pub fn span(&self) -> Span {
+        match *self {
+            QPath::Resolved(_, path) => path.span,
+            QPath::TypeRelative(_, ps) => ps.ident.span,
+            QPath::LangItem(_, span) => span,
+        }
+    }
+
+    pub fn ident_span(&self) -> Span {
```
Comment (petrochenkov) [inline src/librustc_hir/hir.rs:1665]: The name is confusing, this is the span of `()` in `<() as Trait>::method`, why is it an "ident"?
Renaming (`qself_span` or smth) and comments are needed.

File: src/librustc_hir/hir.rs [hunk 2]
```diff
@@ -1677,6 +1681,35 @@ pub enum QPath<'hir> {
...
+        match *self {
+            QPath::Resolved(_, path) => path.span,
+            QPath::TypeRelative(qself, _) => qself.span,
+            QPath::LangItem(_, span) => span,
+        }
+    }
+
+    pub fn sub_span(&self) -> Span {
```
Comment (petrochenkov) [inline src/librustc_hir/hir.rs:1673]: The name is confusing, this is the span of `method` in `<() as Trait>::method`, not sure what "sub" means here.
Renaming (`last_segment_span` or smth) and comments are needed.

File: src/librustc_hir/hir.rs [hunk 3]
```diff
@@ -1493,58 +1493,25 @@ impl Expr<'_> {
...
-            }
-        }
-
-        // `..` desugars to its struct path.
-        ExprKind::Path(QPath::Resolved(None, ref path)) => {
-            return is_range_path(&path) && is_lit(sm, &expr.span);
-        }
+        ExprKind::Struct(ref qpath, _, _) => match **qpath {
+            QPath::LangItem(LangItem::Range, _)
+            | QPath::LangItem(LangItem::RangeTo, _)
```
Comment (petrochenkov) [inline src/librustc_hir/hir.rs:1501]: Could use `QPath::LangItem(LangItem::Range | LangItem::RangeTo | etc... , _)` and `matches!`.

File: src/librustc_passes/lang_items.rs [hunk 4]
```diff
@@ -30,7 +30,13 @@ struct LanguageItemCollector<'tcx> {
 
 impl ItemLikeVisitor<'v> for LanguageItemCollector<'tcx> {
     fn visit_item(&mut self, item: &hir::Item<'_>) {
-        self.check_for_lang(Target::from_item(item), item.hir_id, item.attrs)
+        self.check_for_lang(Target::from_item(item), item.hir_id, item.attrs);
+
+        if let hir::ItemKind::Enum(def, ..) = &item.kind {
+            for variant in def.variants {
+                self.check_for_lang(Target::Variant, variant.id, variant.attrs);
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/librustc_passes/lang_items.rs:37]: [`ItemLikeVisitor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/itemlikevisit/trait.ItemLikeVisitor.html) doesn't have a `visit_variant` function.

File: src/librustc_hir/hir.rs [hunk 5]
```diff
@@ -1444,6 +1447,7 @@ impl Expr<'_> {
             // Partially qualified paths in expressions can only legally
             // refer to associated items which are always rvalues.
             ExprKind::Path(QPath::TypeRelative(..))
+            | ExprKind::Path(QPath::LangItem(..))
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/librustc_hir/hir.rs]: Fixed!

File: src/librustc_typeck/check/mod.rs [hunk 6]
```diff
@@ -123,8 +123,8 @@ use rustc_middle::ty::subst::{self, InternalSubsts, Subst, SubstsRef};
 use rustc_middle::ty::subst::{GenericArgKind, UserSelfTy, UserSubsts};
 use rustc_middle::ty::util::{Discr, IntTypeExt, Representability};
 use rustc_middle::ty::{
-    self, AdtKind, CanonicalUserType, Const, GenericParamDefKind, RegionKind, ToPolyTraitRef,
-    ToPredicate, Ty, TyCtxt, UserType, WithConstness,
+    self, AdtKind, CanonicalUserType, Const, DefIdTree, GenericParamDefKind, RegionKind,
+    ToPolyTraitRef, ToPredicate, Ty, TyCtxt, UserType, WithConstness,
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/librustc_typeck/check/mod.rs]: Fixed!

File: src/librustc_hir/hir.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc_hir/hir.rs]: Fixed!

File: src/librustc_hir/hir.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc_hir/hir.rs]: Fixed!

File: src/librustc_hir/hir.rs (hunk 3 above)
Comment (davidtwco) [reply src/librustc_hir/hir.rs]: Fixed!

**Comment by eddyb** [comment]
> Changes to Clippy might not be correct, they compile but attempting to run tests through ./x.py produced failures which appeared spurious, so I didn't run any clippy tests.

You can use `./x.py test src/tools/clippy` at the very least. Suspecting this is what broke a rollup (#75606), so:
@bors rollup=never

**Comment by eddyb** [comment]
Btw the title still says "Preallocate `DefId`s for lang items" but it doesn't look like that's the strategy taken?
Instead, lang items are referenced by the HIR *without* using their `DefId`s.

(the alternative to this would be doing lang item collection before the HIR, but I think this solution is nicer)

Also, surely this has perf implications?

@bors try @rust-timer queue

**Comment by eddyb** [comment]
Unsurprisingly, this is a nice win for `await`-heavy code.

**Comment by petrochenkov** [comment]
>You can use `./x.py test src/tools/clippy` at the very least.

It's `./x.py test src/tools/clippy --stage 2` now that the defaults were changed.

**Comment by davidtwco** [comment]
Alright - I've ran the clippy tests now and fixed all of the issues that this introduced.

**Comment by ecstatic-morse** [comment]
Final perf results show a [small win](https://perf.rust-lang.org/compare.html?start=30f0a07684f6c1f5df62d69e9519d82e13d6bf2d&end=1656582822a80139d725e56f00c564f4f58f2883&stat=instructions:u) as expected. Nice work!

**Pull request merged**